### DISCUSSION TITLE
OP_RETURN API Support

### DISCUSSION THREAD

**Discussion by B0BSmiths** [CounterpartyXCP/CIPs#132] [General]
CIP: XXX
Title: OP_RETURN API Support
Author: B0BSmith
Discussions-To: ??
Status: Draft
Type: ??
Created: 2024-1-12
 
Abstract

Add the option to only return OP_RETURN data via the API

Rationale

Counterparty mostly runs on OP_RETURN data.  This OP_RETURN data carries encoded Counterparty message data that is then encrypted / obfuscated / encoded using RC4 using the TXID of the first utxo.  It is not currently possible to have the API generate just the OP_RETURN it will only generate fully fledged transactions.  

Counterparty does not have full native P2SH support yet P2SH works fine if you manually create your tx, so rather than having to also manually create the OP_RETURN we could use API to do that for us for a specified utxo

There are tools out there that do help you create the OP_RETURN but if used incorrectly may result in an invalid TX

This will then allow Counterparty to more easily be used with P2SH and P2WSH wallets

Copyright
This document is placed in the public domain.

**Comment by adamkrellenstein** [comment]
Is the alternative just adding proper P2SH support? This feels a bit leaky to me.

**Comment by B0BSmiths** [comment]
Adding p2sh support would be very very nice, would mean we need to send in our redeemScript when making a tx so I felt was more dev heavy than my leaky op return idea that i came up with as a work around to lack of p2sh

The op return idea however does allow more flexibility with tx construction from a power user perspective

**Comment by B0BSmiths** [comment]
If you are to add p2sh support can you please ensure it allows "classic send" where dust can be sent to token recipient, again this is a CPFP thing
